word_conditional_model.py

```diff
@@ -0,0 +1,95 @@
+"""
+    Creating conditional model & predicting  
+"""
+
+import numpy as np
+from nltk.stem import PorterStemmer as stemmer
+import pickle 
+
+class WordConditionalModel:
+
+    def __init__(self):
+        self.__words_matrix = 0
+        self.__bios_matrix = 0
+        self.__model = 0
+        self.__words = []
+    
+    def __build_matrix(self, number_of_sentences, number_of_words):
+        word_matrix = np.zeros((number_of_sentences, number_of_words))
+        return word_matrix
+
+
+    def __represent_sentences_as_matrix(self, matrix, words_lsts):
+        for i in range(len(words_lsts)):
+            for j in range(len(words_lsts[i])):
+                mj = self.__words.index( stemmer().stem(words_lsts[i][j].lower()))
+                for m in range(9):
+                    matrix[(i * 9) + m][mj] = 1
+        return matrix
+
+    def __represent_bios_as_matrix(self, matrix, bios_lsts, words_lsts):
+        for i in range(len(words_lsts)):
+            for j in range(len(words_lsts[i])):
+                d = self.__words.index( stemmer().stem(words_lsts[i][j].lower()))
+                for m in range(9):
+                    if bios_lsts[i][j][m * 2] == "I" or bios_lsts[i][j][m * 2] == "B":
+                        matrix[(i * 9) + m][d] = 1
+        return matrix
+
+    def __calculate_condetional_probs(self, w_i, w_j):
+        occurrenece_of_wi_and_wj, bold_wi = 0, 0
+        occurrenece_of_wi_and_wj = (self.__words_matrix[:, w_i] * self.__words_matrix[:, w_j]).sum()
+        y = self.__words_matrix[:, w_i] * self.__words_matrix[:, w_j]
+        bold_wi = (y * self.__bios_matrix[:, w_i]).sum()
+        self.__model[w_i][w_j] = (bold_wi / occurrenece_of_wi_and_wj) if occurrenece_of_wi_and_wj else 0
+        #print("[{}], [{}] = {}".format(self.__words[w_i], self.__words[w_j], self.__model[w_i][w_j]))
+
+    def fit(self, words_lsts, bios_lsts):
+        print("training WordConditionalModel....")
+        self.__words = []
+        words_vocab = [stemmer().stem(word.lower()) for innerlist in words_lsts for word in innerlist]
+        for word in words_vocab:
+            if word not in self.__words:
+                self.__words.append(word)
+
+        self.__words_matrix = self.__build_matrix(len(words_lsts * 9), len(self.__words))
+        self.__bios_matrix = self.__build_matrix(len(words_lsts * 9), len(self.__words))
+        self.__model = self.__build_matrix(len(self.__words), len(self.__words))
+
+        self.__words_matrix = self.__represent_sentences_as_matrix(self.__words_matrix, words_lsts)
+        self.__bios_matrix = self.__represent_bios_as_matrix(self.__bios_matrix, bios_lsts, words_lsts)
```

Comment:
I think it's better to do not send `self.__words_matrix` as a parameter for these functions. I think moving the `build_matrix` call to the `__represent` function would be nicer.

Reply:
Good point! We will do it.

Reply:
The code has been changed!